in_toto/keylib.go

```diff
@@ -2,220 +2,402 @@ package in_toto
-		return err
+On success it returns the parsed key and nil.
+On failure it returns nil and the error ErrFailedPEMParsing
+*/
+func ParseKey(data []byte) (interface{}, error) {
+	key, err := x509.ParsePKCS8PrivateKey(data)
+	if err == nil {
+		return key, nil
 	}
-
-	return nil
+	key, err = x509.ParsePKCS1PrivateKey(data)
+	if err == nil {
+		return key, nil
+	}
+	key, err = x509.ParsePKIXPublicKey(data)
+	if err == nil {
+		return key, nil
+	}
+	return nil, ErrFailedPEMParsing
 }
 
 /*
-ParseEd25519FromPrivateJSON parses an ed25519 private key from the json string.
-These ed25519 keys have the format as generated using in-toto-keygen:
+LoadKey loads the key file at specified file path into the key object.
+It automatically derives the PEM type and the key type.
+Right now the following PEM types are supported:
 
-	{
-		"keytype: "ed25519",
-		"scheme": "ed25519",
-		"keyid": ...
-		"keyid_hash_algorithms": [...]
-		"keyval": {
-			"public": "..." # 32 bytes
-			"private": "..." # 32 bytes
-		}
-	}
-*/
-func ParseEd25519FromPrivateJSON(JSONString string) (Key, error) {
+	* PKCS1 for private keys
+	* PKCS8	for private keys
+	* PKIX for public keys
+
+The following key types are supported:
+
+	* ed25519
+	* RSA
+	* ecdsa
 
-	var keyObj Key
-	err := json.Unmarshal([]uint8(JSONString), &keyObj)
+On success it will return nil. The following errors can happen:
+
+	* path not found or not readable
+	* no PEM block in the loaded file
+	* no valid PKCS8/PKCS1 private key or PKIX public key
+	* errors while marshalling
+	* unsupported key types
+*/
+func (k *Key) LoadKey(path string, scheme string, keyIdHashAlgorithms []string) error {
+	pemFile, err := os.Open(path)
 	if err != nil {
-		return keyObj, fmt.Errorf("this is not a valid JSON key object")
+		return err
 	}
-
-	if keyObj.KeyType != "ed25519" || keyObj.Scheme != "ed25519" {
-		return keyObj, fmt.Errorf("this doesn't appear to be an ed25519 key")
+	defer func() {
+		if closeErr := pemFile.Close(); closeErr != nil {
+			err = closeErr
+		}
+	}()
+	// Read key bytes and decode PEM
+	pemBytes, err := ioutil.ReadAll(pemFile)
+	if err != nil {
+		return err
 	}
 
-	if keyObj.KeyVal.Private == "" {
-		return keyObj, fmt.Errorf("this key is not a private key")
+	// pam.Decode returns the parsed pem block and a rest.
+	// The rest is everything, that could not be parsed as PEM block.
+	// Therefore we can drop this via using the blank identifier "_"
+	data, _ := pem.Decode(pemBytes)
+	if data == nil {
+		return ErrNoPEMBlock
 	}
 
-	// 64 hexadecimal digits => 32 bytes for the private portion of the key
-	if len(keyObj.KeyVal.Private) != 64 {
-		return keyObj, fmt.Errorf("the private field on this key is malformed")
+	// Try to load private key, if this fails try to load
+	// key as public key
+	key, err := ParseKey(data.Bytes)
+	if err != nil {
+		return err
 	}
 
-	return keyObj, nil
+	// Use type switch to identify the key format
```

Comment:
It would be nice to not hardcode the key type strings here but use constants instead, that are e.g. defined on top of this module. I'm also fine with just a `TODO/FIXME` comment here. :)

Reply:
We should also use those constants in {Generate, Verify}Signature once we have them.

Reply:
done, see newest comment for discussing the exact naming + exporting them or not exporting them.